// react-native-alert-notification/lib/commonjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ALERT_TYPE", {
  enumerable: true,
  get: function () {
    return _ENV.ALERT_TYPE;
  }
});
Object.defineProperty(exports, "Root", {
  enumerable: true,
  get: function () {
    return _containers.Root;
  }
});
Object.defineProperty(exports, "Dialog", {
  enumerable: true,
  get: function () {
    return _containers.Dialog;
  }
});
Object.defineProperty(exports, "Toast", {
  enumerable: true,
  get: function () {
    return _containers.Toast;
  }
});
Object.defineProperty(exports, "IConfigDialog", {
  enumerable: true,
  get: function () {
    return _containers.IConfigDialog;
  }
});
Object.defineProperty(exports, "IConfigToast", {
  enumerable: true,
  get: function () {
    return _containers.IConfigToast;
  }
});
exports.default = void 0;

var _ENV = require("./config/ENV");

var _containers = require("./containers");

var _default = {
  Root: _containers.Root,
  Dialog: _containers.Dialog,
  Toast: _containers.Toast,
  ALERT_TYPE: _ENV.ALERT_TYPE
};
exports.default = _default;
